### ISSUE TITLE
ResNet trained on the full training data (without any splits)

### ISSUE THREAD

**Issue by sayakpaul** [vaishleshik/histology_classifier#2]
@cmacus I have retrained the model using the full training data but it is behaving extremely weird (you can see this in [this notebook](https://github.com/vaishleshik/histology_classifier/blob/master/ResNet_On_Entire_Training_Set.ipynb)). I will debug it tomorrow to find where it is going wrong.

**Comment by magus96** [comment]
Do we have more than 10 images per class available for training?

**Comment by sayakpaul** [comment]
No!
![image](https://user-images.githubusercontent.com/22957388/55825921-86cc5880-5b24-11e9-8e6d-a6e0d2325ee3.png)

I thought of this too. But the loss/accuracy trade-off graph in the mentioned notebook tells a different story.

**Comment by magus96** [comment]
There's another issue, skin3.jpg is 192*256 instead of the other way around which messes up the data generated.

**Comment by magus96** [comment]
I'll upload some notebooks employing different methods to solve this. My inference is that we don't have enough data to teach a model 10 different classes as of now.

**Comment by magus96** [comment]
I noticed you have a loss/accuracy graph there, can you make a notebook with a split dataset? would be easier to find the problem if both training and validation loss is depicted.

**Comment by sayakpaul** [comment]
> I'll upload some notebooks employing different methods to solve this. My inference is that we don't have enough data to teach a model 10 different classes as of now.

Absolutely agreed on this. We have 8 different classes as of now to be precise. 

> I noticed you have a loss/accuracy graph there, can you make a notebook with a split dataset? would be easier to find the problem if both training and validation loss is depicted.

In [this notebook](https://github.com/vaishleshik/histology_classifier/blob/master/Histology_Classification_ResNet.ipynb), I followed the kind of data splitting @cmacus has followed in [his notebook](https://github.com/vaishleshik/histology_classifier/blob/master/histo_keras.ipynb). But there was a problem: there are 8 different classes in the training set and 7 different classes for the validation and test sets. I just saw your notebook and will try using the old faithful **train_test_split ()**. I also liked the model stacking part. I will incorporate these ideas into a separate notebook and share the changes. 

Thank you.

**Comment by cmacus** [comment]
@magus96 please come to my office sometime this week and click some photos for expanding the training data.

**Comment by magus96** [comment]
I've removed skin3.jpg in my fork so you can clone the repo.All images are the same size now so training shouldn't be a problem. Also, In cell 4, you need to enter the subfolder to append the training data to the list. With enough images, I feel it should be able to classify pretty easily.

**Comment by sayakpaul** [comment]
I am pulling that from the 'pool' folder which contains all the training images.

**Comment by cmacus** [comment]
@magus96 @sayakpaul  Great effort guys. The issue really seems to be small training set. Which can be addressed. Feel free to rotate images as you (i.e. skin3.jpeg can be rotated to its proper dimension without making a difference). In a microscope, tissues may come in any spatial orientation, like a cat straight up or inverted.

**Comment by sayakpaul** [comment]
Thank you @sayakpaul. I already incorporating a rotational augmentation in my training loop.

**Comment by sayakpaul** [comment]
@magus96 @cmacus [this notebook](https://github.com/vaishleshik/histology_classifier/blob/master/Sayak's%20Notebooks/Stacked_Models_On_Histologic_Images.ipynb) has pretty good scores on the unseen data (detached from the compute graph). 

Thanks, @magus96. The notebook you pushed was pretty informative and because of that only I further considered to stack ResNet and Logistic Regression (inspired by your idea). As VGG-16, I have fine-tuned the model and it is better than the stacked model.

**Comment by magus96** [comment]
The results seem good. I'll try to implement these with MobileNet and Inception V3 models and compare the performance.

**Comment by sayakpaul** [comment]
Sure. My next step is to try incorporating modern practices like mixed-precision training and one cycle policy. Will update when done.

**Comment by magus96** [comment]
I've never tried using mixed precision for simple ML models. Would be interesting to see the performance boost.

**Comment by cmacus** [comment]
Great job @magus96 and @sayakpaul . What remains is only a validation set now.

**Comment by sayakpaul** [comment]
Yes, primary experiments are done I think.  More sophisticated techniques can be tested on an ad-hoc basis.

**Comment by sayakpaul** [comment]
Yes, primary experiments are done I think.  More sophisticated techniques can be tested on an ad-hoc basis.

**Comment by magus96** [comment]
@sayakpaul @cmacus I have an intuition that the extremely high accuracy here is due to the miniscule val and test data compared to the train data. A random subset from a very small sample size is likely to generate this issue. K fold cross validation and a large test size might shed light. @sayakpaul could you save the weights of the fine tuned VGG16 model? I'll get more images tomorrow and we shall test them on the model.

**Comment by sayakpaul** [comment]
Thanks for the catch @magus96. [Here](https://github.com/vaishleshik/histology_classifier/tree/master/Sayak's%20Notebooks/model_weights), you can find the weights.

**Comment by magus96** [comment]
Could you upload the json file as well? Saves me from recreating the model.

**Comment by sayakpaul** [comment]
I just switched off my laptop and tomorrow I will be travelling for a conference. If I get any chance,  I will surely upload the .json (I myself forgot to serialise the model into JSON). Sorry for inconvenience.

**Comment by magus96** [comment]
No issues, I'll recreate the model in that case.

**Comment by sayakpaul** [comment]
@magus96 I experimented with the mixed precision training today and [this notebook](https://github.com/vaishleshik/histology_classifier/blob/master/Sayak's%20Notebooks/MixedPrecisionTraining.ipynb) contains the experiment. 

Improvements - 
- 10x faster training time
- Accuracy bumped up to _87.5%_ (this still remains under the question and eventually will get validated with a larger set though)

**Comment by magus96** [comment]
I'll upload 50 more images for Lung and Liver class each, let's see if the model works for the binary classification of the the above, then we can extend to other classes.

**Comment by sayakpaul** [comment]
Sure!

**Comment by cmacus** [comment]
Please merge the two folders 'SI' and 'LI' in a single folder called 'intestine'. It makes more sense.

**Comment by sayakpaul** [comment]
Will do. @cmacus

**Comment by cmacus** [comment]
Completed collection of training data. @magus96 please upload all the images in a separate folder 'training_data_jalandhar'. Split some into test images also. We will try all available algorithms and come up with a single notebook comparing different algorithms.

**Comment by sayakpaul** [comment]
Sure.
